src/java/org/apache/cassandra/index/sai/StorageAttachedIndexBuilder.java

```diff
@@ -0,0 +1,351 @@
+
+        CountDownLatch perSSTableFileLock = null;
+        StorageAttachedIndexWriter indexWriter = null;
+
+        Ref<? extends SSTableReader> ref = sstable.tryRef();
+        if (ref == null)
+        {
+            logger.warn(logMessage("Couldn't acquire reference to the SSTable {}. It may have been removed."), sstable.descriptor);
+            return false;
+        }
+
+        try (RandomAccessReader dataFile = sstable.openDataReader();
+             LifecycleTransaction txn = LifecycleTransaction.offline(OperationType.INDEX_BUILD, sstable))
+        {
+            perSSTableFileLock = shouldWritePerSSTableFiles(sstable);
+            // If we were unable to get the per-SSTable file lock it means that the
+            // per-SSTable components are already being built, so we only want to
+            // build the per-index components
+            boolean perIndexComponentsOnly = perSSTableFileLock == null;
+            // remove existing per column index files instead of overwriting
+            IndexDescriptor indexDescriptor = IndexDescriptor.create(sstable);
+            indexes.forEach(index -> indexDescriptor.deleteColumnIndex(index.getIndexContext()));
+
+            indexWriter = new StorageAttachedIndexWriter(indexDescriptor, indexes, txn, perIndexComponentsOnly);
+
+            long previousKeyPosition = 0;
+            indexWriter.begin();
+
+            try (KeyIterator keys = new KeyIterator(sstable.descriptor, metadata))
+            {
+                while (keys.hasNext())
+                {
+                    if (isStopRequested())
+                    {
+                        logger.debug(indexDescriptor.logMessage("Index build has been stopped"));
+                        throw new CompactionInterruptedException(getCompactionInfo());
+                    }
+
+                    final DecoratedKey key = keys.next();
+                    final long keyPosition = keys.getKeyPosition();
+
+                    indexWriter.startPartition(key, keyPosition);
+
+                    RowIndexEntry<?> indexEntry = sstable.getPosition(key, SSTableReader.Operator.EQ);
+                    dataFile.seek(indexEntry.position);
+                    ByteBufferUtil.readWithShortLength(dataFile); // key
+
+                    try (SSTableIdentityIterator partition = SSTableIdentityIterator.create(sstable, dataFile, key))
+                    {
+                        // if the row has statics attached, it has to be indexed separately
+                        if (metadata.hasStaticColumns())
+                            indexWriter.nextUnfilteredCluster(partition.staticRow());
+
+                        while (partition.hasNext())
+                            indexWriter.nextUnfilteredCluster(partition.next());
+                    }
+
+                    bytesProcessed += keyPosition - previousKeyPosition;
+                    previousKeyPosition = keyPosition;
```

Comment:
Can we eliminate `previousKeyPosition` and just do `bytesProcessed != keys.getBytesRead()` right after the loop?

Reply:
Assuming you meant `bytesProcessed += keys.getBytesRead()`.

Reply:
Yup. Typo :D